src/util/caretPositioning.ts

```diff
@@ -0,0 +1,86 @@
+/** These functions use binary search algorithms for efficient caret position calculations within text nodes. */
+import textNodeUtils, { TextNodeLine } from './textNodeUtils'
+
+/**
+ * Performs binary search to find the character offset within a text node
+ * that corresponds to a given X coordinate.
+ *
+ * @param node - The text node to search within.
+ * @param clientX - The X coordinate of the tap/click.
+ * @param lo - Lower bound of the search range (character index).
+ * @param hi - Upper bound of the search range (character index).
+ * @returns The character offset where the caret should be placed.
+ */
+const binarySearchOffset = (node: Text, clientX: number, lo: number, hi: number): number => {
```

Comment:
The name of a function is part of its public API, and should describe what it does, but not how it does it. "Binary search" is an implementation detail, and not something that should be exposed to the caller.